tc/core/tensor-inl.h

```diff
@@ -0,0 +1,322 @@
+  DLConstTensorUPtr res(new DLConstTensor);
+  fillDLTensor(
+      res.get(),
+      DLContext{kDLCPU, 0},
+      tensor.dtype,
+      tensor.shape,
+      tensor.strides,
+      static_cast<const void*>(nullptr),
+      tensor.alignment);
+  return res;
+}
+
+template <typename T>
+inline DLConstTensorUPtr makeDLConstTensor(
+    DLContext ctx,
+    DLDataType dtype,
+    const std::vector<T>& sizes,
+    const std::vector<T>& strides,
+    const void* data,
+    uint64_t byte_offset) {
+  DLConstTensorUPtr res(new DLConstTensor);
+  fillDLTensor(res.get(), ctx, dtype, sizes, strides, data, byte_offset);
+  return res;
+}
+
+// Specializes for const DLTensor*, const DLConstTensor* and TensorInfo
+template <typename T>
+std::vector<DLTensorUPtr> makeDLTensorVector(const std::vector<T>& ptrs) {
+  std::vector<DLTensorUPtr> res;
+  for (auto p : ptrs) {
+    res.push_back(makeDLTensor(p));
+  }
+  return res;
+}
+
+template <typename T>
+std::vector<DLConstTensorUPtr> makeDLConstTensorVector(
+    const std::vector<T>& ptrs) {
+  std::vector<DLConstTensorUPtr> res;
+  res.reserve(ptrs.size());
+  for (auto p : ptrs) {
+    res.push_back(makeDLConstTensor(p));
+  }
+  return res;
+}
+
+template <typename DLTensorPtrType>
+std::vector<TensorInfo> makeTensorInfoVector(
+    const std::vector<DLTensorPtrType>& ts) {
+  std::vector<TensorInfo> iis;
+  iis.reserve(ts.size());
+  std::transform(
+      ts.begin(), ts.end(), std::back_inserter(iis), [](DLTensorPtrType t) {
+        return TensorInfo(t);
+      });
+  return iis;
+}
+
+inline std::vector<const DLTensor*> extractRawPtrs(
+    const std::vector<DLTensorUPtr>& uptrs) {
+  std::vector<const DLTensor*> res(uptrs.size(), nullptr);
+  for (size_t i = 0; i < uptrs.size(); ++i) {
+    res[i] = uptrs[i].get();
+  }
+  return res;
+}
+
+inline std::vector<const DLConstTensor*> extractRawPtrs(
+    const std::vector<DLConstTensorUPtr>& uptrs) {
+  std::vector<const DLConstTensor*> res(uptrs.size(), nullptr);
+  for (size_t i = 0; i < uptrs.size(); ++i) {
+    res[i] = uptrs[i].get();
+  }
+  return res;
+}
+
+inline std::string toString(const DLDataType& t) {
+  if (t.lanes != 1) {
+    CHECK(false) << "NYI: toString for >1 lanes";
+  }
+  switch (t.code) {
+    case DLDataTypeCode::kDLFloat:
+      switch (t.bits) {
+        case 16:
+          return "Half";
+        case 32:
+          return "float";
+        case 64:
+          return "double";
+      }
+      break;
+    case DLDataTypeCode::kDLInt:
+      switch (t.bits) {
+        case 8:
+          return "int8_t";
+        case 16:
+          return "int16_t";
+        case 32:
+          return "int";
+        case 64:
+          return "int64_t";
+      }
+      break;
+    case DLDataTypeCode::kDLUInt:
+      switch (t.bits) {
+        case 8:
+          return "uint8_t";
+      }
+      break;
+  }
+  CHECK(false) << "NYI: toString for type: " << t.code << ", bits: " << t.bits;
```

Comment:
could you use exceptions instead, please?

Reply:
Let's do it once and for all in #351 for the whole codebase, this PR will not start spinning off a solution for #351